specs/opcode/F3RETURN.md

```diff
@@ -0,0 +1,94 @@
+# RETURN opcode
+
+## Procedure
+
+### EVM behavior
+
+The `RETURN` opcode terminates the call successfully with return data for the
+caller.
+
+It behaves differently in different scenarios:
+
+- `is_create` and `is_root`
+    - A. Returns the specified memory chunk as deployment code.
+    - B. End the execution
+- `is_create` and `not is_root`
+    - A. Returns the specified memory chunk as deployment code.
+    - C. Restores caller's context and switch to it.
+- `not is_create` and `is_root`
+    - B. End the execution
+- `not is_create` and `not is_root`
+    - D. Returns the specified memory chunk to the caller.
+    - C. Restores caller's context and switch to it.
+
+### Circuit behavior
+
+The circuit first checks the `result` in call context is indeed success.
+
+We define source memory chunk by the `RETURN` arguments `return_offset` and
+`return_length`, obtained from the stack via 2 lookups.
+
+Perform memory expansion to `return_offset + return_length`.
+
+A. If it's an `is_create` call context, it copies the source memory chunk to
+the bytecode identified by its hash, using a lookup to the copy circuit with
+the following parameters:
+    - `src_id: callee.call_id`
+    - `src_type: Memory`
+    - `src_addr: return_offset`
+    - `src_addr_end: return_offset + return_length`
+    - `length: return_length`
+    - `dst_id: code_hash`
+    - `dst_type: Bytecode`
+    - `dst_addr: 0`
+    - `rw_counter: callee.rw_counter`
+    - `rwc_inc: length`
+D. Otherwise, if it's not a root call, it copies the source memory chunk to the
+callers memory defined by the `*CALL*` arguments `retOffset`, `retLength`,
+using a lookup to the copy circuit with the following parameters:
+    - `src_id: callee.call_id`
+    - `src_type: Memory`
+    - `src_addr: return_offset`
+    - `src_addr_end: min(return_offset + return_length, callee.memory_size)`
```

Comment:
Right!  The python part doesn't do the min.  And since there's memory expansion, it's OK to read past the memory_size (and the state circuit only allows 0s to be read in such case).

Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/229/commits/c73108e99f2d0e674a3d758f183aa82bc21cf95c